Create a SuperPMI collection pipeline

BruceForstall opened the pull request:
Update from Kunal:

Added `runtime superpmi` pipeline that does the following:

- Builds CoreClr/Libraries/Core_Root
- Runs SuperPMI setup script that:
  - Clones and builds jitutils to produce pmi.dll.
  - Prepare correlation payload directory. This is the folder that contains Core_Root, pmi.dll, superpmi scripts that is needed to run SPMI collection. This is a common directory that is sent to all the helix machines that will do the collection.
  - Scans the folder on which SPMI need to be done (in this case, it is just Core_Root, but in future will be TestArtifacts) and partition the .dlls/.exe in chunks of 50MB. These partition constitutes “payload” directory that is sent to individual helix machines to do the collection in parallel. For e.g. Core_Root folder is 150 MB (just managed .dll and  .exe) and is divided in 3 partitions so we can do SPMI collection on 3 machines in parallel, each doing it on payload of 50 MB.
- Sends payload/correlation directories to helix machine and performs the SPMI collection.
- Once helix machines are done with collection, there will be multiple .mch files produced from each partition (libraries.0.mch, libraries.1.mch, libraries.2.mch in this case). We bring them back on azure machine and merge and de-duplicate all the .mch files into a single libraries.mch file.
- Upload the merged libraries.mch file to azure blob storage.

I also had to update `superpmi.py` to make it Python3.5 compliant. I didn't remove the test run part but commented the portion of tests collection with TODO intentionally so when I have to enable it, I will know what to enable.

BruceForstall committed:
Create a SuperPMI collection pipeline

Start with a clone of jitstress.yml, and go from there.

kunalspathak committed:
Only build Windows_NT_x64, disable test builds and runs

BruceForstall committed:
TEMPORARY: disable normal PR triggered jobs

kunalspathak committed:
some template files for superpmi

kunalspathak committed:
Little trial error# 1

kunalspathak committed:
fix the superpmi-setup.ps1

kunalspathak committed:
Disable CoreClrTestBuildHost, fix the project path in superpmi-send-to-helix

kunalspathak committed:
Skip build, download, create core_root

kunalspathak committed:
Just copy superpmi scripts, import pipeline-logging-functions

kunalspathak committed:
add a dummy step

kunalspathak committed:
add missing space before message

kunalspathak committed:
Replace message with script

kunalspathak committed:
Temporarily disable dependsOn

kunalspathak committed:
disable another dependsOn

kunalspathak committed:
Pass superpmi.proj file

kunalspathak committed:
logging around send-to-helix

kunalspathak committed:
fix superpmi.proj to include ItemGroup

kunalspathak committed:
have placeholder.txt in WorkItem folder

kunalspathak committed:
fix workitem folder creation, script name

kunalspathak committed:
Re-enable build coreclr, libraries, download, core_root

kunalspathak committed:
fix superpmi.py path

kunalspathak committed:
clone jitutils, build and copy pmi.dll

kunalspathak committed:
Pass Core_Root as payload

kunalspathak committed:
Use hardcoded checked

kunalspathak committed:
Use .dotnet from runtime repo
- Build jitutils before and send the payload to helix

kunalspathak committed:
Fix the pmi.dll location and add  -output_mch_path

kunalspathak committed:
fix the output path

kunalspathak committed:
Try publishing SourcesDirectory

kunalspathak committed:
Upload to HELIX_WORKITEM_UPLOAD_ROOT

kunalspathak committed:
Include test in collection and partition core_root and test

kunalspathak committed:
Enable CoreClrTestBuildHost

kunalspathak committed:
Add dependsOn P1 test

kunalspathak committed:
Add DownloadFilesFromResults

kunalspathak committed:
Properly copy the Tests

kunalspathak committed:
Workaround for https://github.com/dotnet/arcade/pull/6179

kunalspathak committed:
Disable test run, SuperPMI just system.web.dll

kunalspathak committed:
Fix the mch path

kunalspathak committed:
Comment partition for test

MattGal commented on eng/common/templates/steps/superpmi-send-to-helix.yml:
```diff
@@ -0,0 +1,68 @@
+      DotNetCliVersion: ${{ parameters.DotNetCliVersion }}
+      EnableXUnitReporter: ${{ parameters.EnableXUnitReporter }}
+      WaitForWorkItemCompletion: ${{ parameters.WaitForWorkItemCompletion }}
+      Creator: ${{ parameters.Creator }}
```
Creator is for public / open scenarios and HelixAccessToken is for authenticated ones.  You need to avoid specifying Creator in any case where it's public as it will be inferred from the token.

MattGal replied:
Also this is reaaaaly close to https://github.com/dotnet/arcade/blob/master/eng/common/templates/steps/send-to-helix.yml and should probably just be using that.

kunalspathak replied:
I can't use `send-to-helix.yml` because it invokes hardcoded `helixpublish.proj` while I want to invoke a different `proj` file. I can certainly update `send-to-helix.yml` to take the `ProjectFile` as parameter, but I will leave it for different PR.

kunalspathak committed:
Remove commented code

kunalspathak committed:
Increase timeout to 12 hours

kunalspathak committed:
fast pass

- Comment build coreclr/libraries job
- Instead just download coreclr from some previous run, and tell that it is core_root

kunalspathak committed:
Update the machine pool

kunalspathak committed:
Add a dummy step

kunalspathak committed:
Use SuperPMI_Min.zip that contains all the required files

kunalspathak committed:
Another way of Extracting cab

kunalspathak committed:
download in powershell, extract files separately

kunalspathak committed:
fix the zip path

kunalspathak committed:
Another logic of download zip

kunalspathak committed:
fix path

kunalspathak committed:
fix newdir path

kunalspathak committed:
more fixes

kunalspathak committed:
fix extract step

kunalspathak committed:
add test python script and modify creator

kunalspathak committed:
Map secret to env

kunalspathak committed:
decrease timeout, fix superpmi path in proj, print unzipped contents

kunalspathak committed:
Add a step to download and dir artifacts

kunalspathak committed:
Attempt to call super_pmi upload

kunalspathak committed:
list all contents

kunalspathak committed:
Also dir SourcesDir\artifacts

kunalspathak committed:
Use DownloadPipelineArtifact

kunalspathak committed:
Use the right key

kunalspathak committed:
Upload from helixresults

kunalspathak committed:
cleanup run-superpmi-job.yml

kunalspathak committed:
Re-enable superpmi for CORE_ROOT

kunalspathak committed:
fix the dependsOn

kunalspathak committed:
Replace checked with $(BuildConfig)

kunalspathak committed:
Replace checked with $(BuildConfig)
Comment download artifacts

kunalspathak committed:
change helix queue for windows and add for arm64

kunalspathak committed:
Use FileSeparateChar in superpmi.proj

kunalspathak committed:
Add Linux x64

kunalspathak committed:
Fix the helix queue, property erroro

kunalspathak committed:
logging in shell

kunalspathak committed:
Fix PayloadDir and CorrelationPayloadDir

kunalspathak committed:
Make timeout to 5 hours

kunalspathak committed:
Print workitem contents

kunalspathak committed:
fix WorkItemDirectory

kunalspathak committed:
Just copy pmi.dll

kunalspathak committed:
fix the pmi_assemblies path

kunalspathak committed:
disable reporter

kunalspathak committed:
pop put JitUtilsDirectory before deleting it

kunalspathak committed:
chmod for superpmi-setup.sh

kunalspathak committed:
just copy pmi.dll for unix

kunalspathak committed:
fix the parameter to unix

kunalspathak committed:
Add logging to understand -pmi_assemblies issue

kunalspathak committed:
Add "binaries" folder where payload is present
fix unit script that initializes array

kunalspathak committed:
remove precommand

kunalspathak committed:
Pass Core_Root as CorrelationPayload

- Pass Core_Root as CorrelationPayload
- Pass tag to name .mch file with
- Changes to .ps1 and .sh to copy Core_Root directly

kunalspathak committed:
Add back the workaround for https://github.com/dotnet/arcade/pull/6179

Also, rename 'checked' -> 'Checked'

kunalspathak committed:
fix checked for Linux

kunalspathak committed:
mkdir for linux

kunalspathak committed:
fix path in HelixResultsDestinationDir

kunalspathak committed:
Enable Windows x86/arm64

kunalspathak committed:
fix the python path

kunalspathak committed:
First version of superpmi-setup.py

kunalspathak committed:
Attempt to use superpmi-setup.py

kunalspathak committed:
Fix spelling of LibrariesArtifacts, set python path

kunalspathak committed:
TestArtifacts commented

kunalspathak committed:
superpmi-setup script for linux

kunalspathak committed:
Enable Linux_x64

kunalspathak committed:
Added pydocs

kunalspathak committed:
Add option for exclude_directories

kunalspathak committed:
Moved all the logic to superpmi-setup.py

kunalspathak committed:
Replace superpmi-setup.ps1 and superpmi-setup.sh with superpmi-setup.py

kunalspathak committed:
Fix up in superpmi-setup.py

kunalspathak committed:
fix display name

kunalspathak committed:
add missing spacE

kunalspathak committed:
Use differnet conditional

kunalspathak committed:
ignore PermissionError while cleaning up temp folder

kunalspathak committed:
fix workitem copy for linux

kunalspathak committed:
fix copy for unix

kunalspathak committed:
Publish logs

kunalspathak committed:
Print LibrariesArtifacts

kunalspathak committed:
remove space in front of publish logs desc

kunalspathak committed:
Add missing quotes for LibrariesArtifacts

kunalspathak committed:
Download x64 coreclr to make upload work

kunalspathak committed:
Enable Windows x86/arm64, Linux arm/arm64

kunalspathak committed:
Delete superpmi-setup.ps1, superpmi-setup.sh

kunalspathak committed:
Fix space around LibrariesArtifacts

kunalspathak committed:
Download real coreclr x64

kunalspathak committed:
ignore native dlls

kunalspathak committed:
Make superpmi.py Python3.6 friendly

kunalspathak committed:
Minor cleanup

kunalspathak committed:
Cleanup superpmi-send-to-helix to use helix-inner-step

kunalspathak committed:
Remove redundant steps

kunalspathak committed:
Merge mch files

kunalspathak committed:
Pass osGroup

kunalspathak committed:
Make superpmi-setup script takes tests_directory

kunalspathak committed:
Enable SPMI collection for P1

kunalspathak committed:
General cleanup

kunalspathak committed:
fix merge_mch pattern for OS

kunalspathak committed:
Log the SPMI collection output and put it in upload_root folder

kunalspathak committed:
Replace robocopy/rsync with shutil.copy2

Note that shutil.copy2 is very slow specially on Linux as compared to rsync.
But I think we are ok to have slowness in return of simpler code.

kunalspathak committed:
fix path of pmi.dll to be copied

kunalspathak committed:
Include linux .so and file_exe files in copy_directories

kunalspathak committed:
Try publishing just .mch file

kunalspathak committed:
Revert "Try publishing just .mch file"

This reverts commit a5312c05f9502c7c2f2c5a28125f75f8d824b869.

kunalspathak committed:
Disable tests SPMI collection on 42053

- Tests run times out on Windows
- Some of the partition's collection failed for tests because reply was not clean
- Here are the details when libraries/tests were run:
  https://dev.azure.com/dnceng/internal/_build/results?buildId=832814&view=results
- Hence, disable tests run and just have libraries run for now.
- I didn't remove the test run part but commented the portion of tests collection with
  TODO intentionally so when I have to enable it, I will know what to enable.

kunalspathak commented:
@dotnet/jit-contrib

kunalspathak commented:
@safern - Thanks a lot for helping me through this. If you get a chance, please take a look at the overall pipeline structure.

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+#
+# Notes:
+#
+# Script to setup directory structure to perform SuperPMI collection in CI.
```
It would be useful for this top comment to be more descriptive: what the inputs, outputs, what does it need to do to "setup directory structure..."?

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+parser.add_argument("-tests_directory", help="path to managed test artifacts directory")
+parser.add_argument("-max_size", help="Max size of each partition in MB")
+is_windows = platform.system() == "Windows"
+native_binaries_to_ignore = [
```
Presumably there needs to be a Linux and macOS version of this list

kunalspathak replied:
Yes, I have added it in my TODO list of #42783 .

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+    coreclr_args = setup_args(main_args)
+    source_directory = coreclr_args.source_directory
+
+    # CorelationPayload directories
```
typo: Corelation

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+    superpmi_src_directory = path.join(source_directory, 'src', 'coreclr', 'scripts')
+    superpmi_dst_directory = path.join(correlation_payload_directory, "superpmi")
+    arch = coreclr_args.arch
+    helix_source_prefix = "official"
```
What does "official" mean?

kunalspathak replied:
I am not entirely sure. I believe we need to set the `HelixSource` to a value that starts with `official`.

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+    helix_source_prefix = "official"
+    creator = ""
+    ci = True
+    if is_windows:
```
I'm worried about hard-coding these Helix queues into a Python script in src\coreclr\scripts. Who will remember these are here, versus in the YML files in the `eng` directory, when they need to be updated? Could they at least be put in some kind of table at the top of the file so they are more visible?

kunalspathak replied:
For now I will keep it here and include the queue in the table at the top of the file as you suggested. May be in next iteration, I will move this file to `eng\common\superpmi\`. That is where other related files are present.

BruceForstall commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -0,0 +1,395 @@
+                ["git", "clone", "--quiet", "--depth", "1", "https://github.com/dotnet/jitutils", jitutils_directory])
+            # Set dotnet path to run bootstrap
+            os.environ["PATH"] = path.join(source_directory, ".dotnet") + os.pathsep + os.environ["PATH"]
+            if is_windows:
```
Might be better to set the bootstrap string name, then only have one copy of the `run_command` line

BruceForstall reviewed (commented):
Looks good.

(I can't approve this because it's technically "my" PR)

kunalspathak commented:
> Looks good.
> 
> (I can't approve this because it's technically "my" PR)

Thanks Bruce for reviewing. I will incorporate the feedback and approve it! :)

kunalspathak commented on eng/pipelines/coreclr/superpmi.yml:
```diff
@@ -0,0 +1,66 @@
+
+pr: none
+
+# schedules:
```
Sure, scheduled it to run every Sunday at 9AM PST.

kunalspathak commented on eng/pipelines/coreclr/templates/run-superpmi-job.yml:
```diff
@@ -0,0 +1,143 @@
+
+    # For now, we won't upload merged collection as an artifact.
+
+    - script: $(PythonScript) $(Build.SourcesDirectory)/src/coreclr/scripts/superpmi.py upload -arch $(archType) -build_type $(buildConfig) -mch_files $(MchFilesLocation)libraries.pmi.$(MchFileTag).mch -core_root $(Build.SourcesDirectory)/artifacts/bin/coreclr/$(osGroup).x64.$(buildConfigUpper)
```
Yes, I am doing that in #42783

kunalspathak commented on eng/pipelines/coreclr/templates/superpmi-job.yml:
```diff
@@ -0,0 +1,104 @@
+    # Create Core_Root
+    - script: $(Build.SourcesDirectory)/src/tests/build$(scriptExt) $(buildConfig) $(archType) generatelayoutonly $(librariesOverrideArg)
+      displayName: Create Core_Root
+      condition: and(succeeded(), ne(variables.runtimeFlavorName, 'Mono'))
```
Yes, removed it.

kunalspathak commented on eng/common/superpmi/superpmi.proj:
```diff
@@ -0,0 +1,103 @@
+    <EnableAzurePipelinesReporter>false</EnableAzurePipelinesReporter>
+    <EnableXUnitReporter>false</EnableXUnitReporter>
+    <WorkItemTimeout>5:00</WorkItemTimeout>
+    <WorkItemTimeout Condition="'$(HelixSourcePrefix)' != 'official'">0:15</WorkItemTimeout>
```
I believe we need to set `HelixSourcePrefix` to something that starts with `official/` or `pr/`, but I don't think we need this condition because I have set it to `official`. I will remove this condition.

kunalspathak committed:
Review feedback

kunalspathak reviewed (approved):
LGTM

safern commented on eng/common/superpmi/superpmi.proj:
```diff
@@ -0,0 +1,102 @@
+    <FileSeparatorChar>\</FileSeparatorChar>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(AGENT_OS)' != 'Windows_NT'">
+    <FileSeparatorChar>/</FileSeparatorChar>
```
You don't need this. You can use `MSBuild::NormalizeDirectory` or `MSBuild::NormalizePath` apis. We have a bunch of usages and examples in the runtime repo.

safern replied:
Note that this APIs need to be used with full paths, it doesn't work with relative paths.

safern replied:
Maybe you do need it... because you're using it to calculate the commands you're going to run on helix.

safern commented on eng/pipelines/coreclr/templates/run-superpmi-job.yml:
```diff
@@ -0,0 +1,143 @@
+      # Run superpmi collection in helix
+    - template: /eng/common/templates/steps/superpmi-send-to-helix.yml
+      parameters:
+        HelixSource: '$(HelixSourcePrefix)/$(Build.Repository.Name)/$(Build.SourceBranch)' # sources must start with pr/, official/, prodcon/, or agent/
```
I beliueve this is no longer needed, this is automatically set by the Helix SDK nowadays.

safern commented on eng/pipelines/coreclr/templates/run-superpmi-job.yml:
```diff
@@ -0,0 +1,143 @@
+    - script: $(PythonScript) $(Build.SourcesDirectory)/src/coreclr/scripts/superpmi.py upload -arch $(archType) -build_type $(buildConfig) -mch_files $(MchFilesLocation)libraries.pmi.$(MchFileTag).mch -core_root $(Build.SourcesDirectory)/artifacts/bin/coreclr/$(osGroup).x64.$(buildConfigUpper)
+      displayName: Upload SuperPMI libraries collection to Azure Storage
+      env:
+        CLRJIT_AZ_KEY: $(clrjit_key1) # secret key stored as variable in pipeline
```
We usually try to avoid key tokens in public pipelines. Is this pipeline public?

BruceForstall replied:
The pipeline is internal, precisely due to this requirement.
